Changes required to integrate the backup

ExcChua opened the pull request:
Terraform configuration changes/merges required to integrate the backup. Fixes #16

ExcChua committed:
Changes required to integrate the backup

Changes required to integrate the backup

Lsquared13 reviewed (changes requested):
I'm having trouble picturing how this would look fully integrated just by looking at the code. I also think all the installation scripts are unnecessary clutter and that we wouldn't really be able to make use of them in a TFE-based workflow anyway.

Can you please get rid of the extra scripts and create a PR that already has everything fully integrated? You can create new terraform config files for organizational purposes (not including a new variable or output file), but place them directly into the directory they belong in. You can also modify existing files if you think it's appropriate.

Essentially I want you to be able to do the standard `packer build`, `copy-packer-artifacts-to-terraform.py`, and `terraform apply` workflow without any extra steps and have it launch a network that has backup functionality enabled. Let me know if you have any questions.

ExcChua committed:
Integrated backup contents directly into existing files

Integrated backup contents directly into existing files

ExcChua commented:
Hi Louis,

Changes made as requested.

Lsquared13 commented:
Thanks, I still see a bunch of files in the BackupLambda-terraform directory. Maybe you need to finish cleaning it up? I'll take a look at the other stuff but please remove that directory from the PR when you can.

Lsquared13 commented on terraform/main.tf:
```diff
@@ -4,6 +4,13 @@ module "quorum_cluster" {
 
   # Variables sourced from terraform.tfvars
   public_key_path                    = "${var.public_key_path}"
+  private_key_path                   = "${var.private_key_path}"
```
This is for the private key that BackupLambda will use to SSH the machines right? The one we set up as a FoxPass service account?

If that is the case lets name it something else please. Something maybe like `backup_lambda_ssh_private_key_path`. I generally assume `private_key_path` corresponds to the same key as `public_key_path` and it seems like that's not actually the case here.

If it does correspond to the same key we initially launch the network with, then why do we need it now?

ExcChua replied:
> This is for the private key that BackupLambda will use to SSH the machines right? The one we set up as a FoxPass service account?

Yes, it is.

> If it does correspond to the same key we initially launch the network with, then why do we need it now?

I'm confused. Your question seems to imply it's not needed. How would the BackupLambda connect to the nodes then?

Lsquared13 replied:
If it's the backup service account key then it's not the same key we initially launch the network with and I understand why we need it. Given this I just think the name should be changed.

Lsquared13 commented on terraform/main.tf:
```diff
@@ -4,6 +4,13 @@ module "quorum_cluster" {
+  private_key_path                   = "${var.private_key_path}"
+  backup_interval                    = "${var.backup_interval}"
+  BackupLambda_source_file           = "${var.BackupLambda_source_file}"
+  BackupLambda_output_path           = "${var.BackupLambda_output_path}"
```
nit: Let's use `backup_lambda` rather than `BackupLambda` for these. I understand you were shooting for consistency with the name of the application but I'd rather not introduce camel case into the Terraform configs.

ExcChua replied:
These are now using backup_lambda...

Lsquared13 commented on terraform/modules/quorum-cluster-region/outputs.tf:
```diff
@@ -37,3 +37,7 @@ output "quorum_validator_cidr_block" {
   value = "${lookup(var.observer_node_counts, var.aws_region, 0) > 0 ? data.template_file.quorum_observer_cidr_block.rendered : ""}"
 }
+
+output "quorum_cluster_main_route_table_id" {
```
nit: maybe `quorum_vpc_main_route_table_id` or `region_main_route_table_id` instead here? The route table doesn't really apply to the whole cluster as this would seem to imply.

ExcChua replied:
Renamed to quorum_vpc_main_route_table_id

ExcChua replied:
This has been renamed to quorum_vpc_main_route_table_id

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+
+# This is to be placed in quorum-cluster-region
```
nit: unnecessary comment

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+
+# This is to be placed in quorum-cluster-region
+
+provider "archive" {}
```
nit: add version strings to these providers. I'm pretty sure it tells you exactly what strings to add when you run `terraform init`

ExcChua replied:
Version strings added

ExcChua replied:
Version added

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+
+provider "archive" {}
+provider "random" {}
+provider "null" {}
```
nit: Double check that you need the null provider here and can't just inherit it. I know I've used it but it may or may not be in this module.

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+provider "random" {}
+provider "null" {}
+
+data "local_file" "private_key" {
```
This will break TFE because the file won't exist. You need to add the option to provide the key directly in the variables file. If you keep the option to load it from a file there will need to be a `count` variable that ensures this data source is not created if the key is passed directly. I've dealt with this before so you should be able to look to my handling of SSH keys for examples.

ExcChua replied:
Fixed

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+  filename = "${var.private_key_path}"
+}
+
+resource "aws_s3_bucket_object" "object" {
```
Can this name be more descriptive than `object`?

ExcChua replied:
fixed

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+  source = "${var.enc_ssh_path}-${var.aws_region}"
+}
+
+resource "aws_sns_topic" "BackupEvent" {
```
nit: Applies across this whole page but again, please no camel case in resource names. Using it in the variables like you do with `name_prefix` is fine.

ExcChua replied:
fixed

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+}
+
+
+resource "aws_security_group_rule" "quorum_maker_https" {
```
Is this rule necessary? I'm pretty sure they already have egress open everywhere on all ports.

ExcChua replied:
fixed

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+}
+
+resource "aws_cloudwatch_event_rule" "BackupTimer" {
+  count = "${signum(lookup(var.maker_node_counts, var.aws_region, 0))}"
```
nit: leave an empty line after `count` variables

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+# Allow the SNS to trigger the backup lambda
+resource "aws_lambda_permission" "BackupLambda" {
+  count = "${signum(lookup(var.maker_node_counts, var.aws_region, 0))}"
+  statement_id = "AllowExecutionFromSNS-BackupLambda-${var.network_id}-${var.aws_region}"
```
nit: this applies to the whole page at least. Align your `=` horizontally among variable declarations that are on adjacent lines.

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+   policy_arn = "${aws_iam_policy.BackupLambda_logging.arn}"
+}
+
+# Zip the Backup Lambda
```
Per my earlier email I think we should be downloading already-zipped releases from GitHub

ExcChua replied:
Now retrieving zip from GitHub

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+  depends_on = ["null_resource.BackupLambda"]
+}
+
+resource "aws_kms_key" "a" {
```
Please use a more descriptive name for this

ExcChua replied:
Renamed

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+  }
+}
+
+output "aws_kms_key_test" {
```
Outputs should not be mixed with the rest of the configuration. Also, this looks like maybe it was left around when it shouldn't be?

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+  plaintext = "${file("${var.private_key_path}")}"
+}
+
+# Save the encrypted contents to the file specified at filename
```
What are you getting out of saving the file? We can't expect it to stick around on TFE

Lsquared13 replied:
**bump** this comment

Lsquared13 replied:
Please answer this question, I do want to understand the purpose of this resource.

ExcChua replied:
This encrypts the SSH private key for the BackupLambda function to SSH to the nodes

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+  filename = "${var.enc_ssh_path}-${var.aws_region}"
+}
+
+resource "aws_kms_grant" "a" {
```
More descriptive name please

ExcChua replied:
fixed

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+  operations        = ["Encrypt", "Decrypt"]
+}
+
+output "aws_s3_bucket" {
```
Again outputs shouldn't be mixed with other config. Also this name is far too general for an output name in this module anyway. We need to specify its purpose.

ExcChua replied:
This is used for debugging

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+  value = "${element(concat(aws_s3_bucket.quorum_backup.*.id, list("")), 0)}"
+}
+
+resource "aws_security_group" "allow_all" {
```
This seems like a scary security group. What is it used for?

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+    description = "Allow all outgoing traffic"
+    vpc_id = "${aws_vpc.quorum_cluster.id}"
+
+    egress {
```
I prefer to use `aws_security_group_rule` resources rather than specifying rules inline

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+}
+
+// use the next value after data.template_file.quorum_observer_cidr_block
+data "template_file" "quorum_maker_cidr_block_lambda" {
```
If this is just for the lambda we don't need a big network. Maybe we should split this CIDR into a few subnets and just assign this to the first one?

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+  owners = ["099720109477"] # Canonical
+}
+
+resource "aws_instance" "web1" {
```
More descriptive name please. I don't actually know what this instance does.

ExcChua replied:
The entire Terraform configuration was developed modularly, in isolation from each other. This was used during development to ensure that instances can be reached and connected via SSH, to test VPC, and to test the BackupLambda subnet.

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,399 @@
+  }
+}
+
+resource "aws_eip" "gw1" {
```
more descriptive name please

ExcChua replied:
gw1 -> gateway 1

Lsquared13 commented on terraform/modules/quorum-cluster-region/variables.tf:
```diff
@@ -382,3 +382,51 @@ variable "observer_node_counts" {
+# If singular, it's 1 hour, 1 minute, 1 week, 1 day, etc.
+variable "backup_interval" {
+    default = ""
+    description = "The scheduled expression for the backup interval"
```
Put a few examples in the description please

ExcChua replied:
Examples done

Lsquared13 commented on terraform/modules/quorum-cluster-region/variables.tf:
```diff
@@ -382,3 +382,51 @@ variable "observer_node_counts" {
+
+// this creates an AWS test instance for the Lambda to SSH into
+// with the KMS encrypted key, which it will decrypt within memory
+variable "debug" {default=""}
```
Maybe get rid of this? If you think it's helpful, format it like the others.

ExcChua replied:
Removed

Lsquared13 commented on terraform/outputs.tf:
```diff
@@ -17,3 +17,11 @@ output "bootnode_ips" {
   value = "${module.quorum_cluster.vault_server_ips}"
 }
+
+output "private_key_path" {
```
nit: These outputs are the same

Lsquared13 reviewed (changes requested):
I'm not sure how reasonable it would be in terms of dependency order but if it's plausible to make it a separate module (in `terraform/modules`) that each `quorum-cluster-region` uses that might be ideal. You don't have to do it if it seems too clunky.

ExcChua committed:
Removed BackupLambda-terraform directory

Removed BackupLambda-terraform directory as the functionality has been integrated

Lsquared13 commented:
Also, regarding the NAT Gateway vs. NAT Instance discussion, let's just go with the gateway as a managed service.

ExcChua committed:
Changes as requested by Louis

Changes as requested by Louis

ExcChua committed:
Update and force run

Update and force run

ExcChua committed:
Changed data.archive_file to null_resource.zip_backup_lambda

Changed how the backup lambda is zipped.

ExcChua committed:
Fix for fetching backup lambda and Terraform destroying temp files

Fix for Terraform destroying any temp files

ExcChua committed:
Fix backup lambda output

Fix backup lambda output

Lsquared13 commented on terraform/modules/quorum-cluster-region/outputs.tf:
```diff
@@ -37,3 +37,7 @@ output "quorum_validator_cidr_block" {
 }
+
+output "quorum_vpc_main_route_table_id" {
+  value = "${length(aws_vpc.quorum_cluster.*.id) != 0 ? element(concat(aws_vpc.quorum_cluster.*.main_route_table_id, list("")), 0) : ""}"
```
I think this can be simplified to `value = "${element(coalescelist(aws_vpc.quorum_cluster.*.main_route_table_id, list("")), 0)}"`

coalescelist seems a little cleaner to me for avoiding the empty list issue (though admittedly I only discovered it somewhat recently) and you don't need the conditional since you'd just be grabbing an empty string out of the list if the VPC doesn't exist.

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,329 @@
+}
+
+resource "aws_s3_bucket_object" "encrypted_ssh_key" {
+  count      = "${signum(lookup(var.maker_node_counts, var.aws_region, 0))}"
```
style nit: empty line after `count` variables

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,329 @@
+
+resource "aws_s3_bucket_object" "encrypted_ssh_key" {
+  count      = "${signum(lookup(var.maker_node_counts, var.aws_region, 0))}"
+  depends_on = ["data.aws_kms_ciphertext.encrypt_ssh_operation", "local_file.EncryptedSSHKey"]
```
style nit: put `depends_on` at the end of the resource with an empty line above

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,329 @@
+}
+
+resource "aws_sns_topic" "backup_event" {
+  count       = "${signum(lookup(var.maker_node_counts, var.aws_region, 0))}"
```
Why is the count related only to the maker nodes in the region? Would we not want the backup to run in a region with only validators or observers?

I suspect we want these resources if there are any nodes of any type, but if they're going to be only attached to one type of node I'd also suggest it should be validators. Let me know what the intent is here.

ExcChua replied:
As discussed, backup 1 (randomly) of validators if it exist. If not, then 1 (randomly) of maker, and if not, then 1 (randomly) of observers.

Lsquared13 replied:
FWIW it doesn't actually need to be random, just arbitrary. May as well do it randomly if that's easy though.

Lsquared13 replied:
**bump** this comment

I know there's Go code too but if you made this change you'll need to change the `count` variables, right?

ExcChua replied:
Arbitrarily in computing terms means random, unless you're talking about me choosing a fixed value, in that case, it's called fixed.

And please provide a style guide so I can fix everything all at once.

ExcChua replied:
Your instructions were to fix existing changes first, then fix the count. Unless your instructions are suggestions, in which case, I would then ignore it.

So, which is it? Instructions or suggestions?

Lsquared13 commented on terraform/modules/quorum-cluster-region/variables.tf:
```diff
@@ -98,6 +98,20 @@ DESCRIPTION
 
+variable "backup_lambda_ssh_private_key_path" {
+  description = <<DESCRIPTION
+SSH private key to be used for authentication.
```
nit: **Path to** SSH private key to be used for authentication **by the BackupLambda function**

ExcChua replied:
Fixed

Lsquared13 commented on terraform/modules/quorum-cluster-region/variables.tf:
```diff
@@ -98,6 +98,20 @@ DESCRIPTION
+
+variable "backup_lambda_ssh_private_key" {
+  description = <<DESCRIPTION
+SSH private key to be used for authentication.
```
nit: SSH private key to be used for authentication **by the BackupLambda function**

Lsquared13 commented on terraform/modules/quorum-cluster-region/variables.tf:
```diff
@@ -382,3 +396,59 @@ variable "observer_node_counts" {
 }
+
+variable "backup_lambda_binary" {
+    default = "BackupLambda"
```
style nit: put `description` before `default`

Lsquared13 replied:
**bump** this comment

Lsquared13 replied:
style nit: Please address this and put `description` above `default`

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,329 @@
+  provisioner "local-exec" {
+     command = <<EOT
+ if [ ! -e ${var.aws_region}-${var.backup_lambda_output_path} ]; then 
+   wget -O ${var.aws_region}-${var.backup_lambda_output_path} https://github.com/EximChua/BackupLambda/releases/download/0.1/BackupLambda.zip
```
If you have a URL parameter shouldn't you be using it here?

ExcChua replied:
Fixed

Lsquared13 commented on terraform/modules/quorum-cluster/variables.tf:
```diff
@@ -460,3 +476,43 @@ variable "observer_node_counts" {
+  default = ""
+}
+
+variable "backup_lambda_binary" {
```
I'm searching through the PR and I can't find where this `backup_lambda_binary` variable is used. Please either point out where it is used, modify some hard-coded things to use this variable instead, or remove this variable completely.

ExcChua replied:
Used in definition of aws_lambda_function. Fixed.

Lsquared13 commented on terraform/modules/quorum-cluster/variables.tf:
```diff
@@ -460,3 +476,43 @@ variable "observer_node_counts" {
+  description = "Name of BackupLambda binary"
+}
+
+variable "backup_lambda_binary_url" {
```
I'm searching through the PR and I can't find where this `backup_lambda_binary_url` variable is used. Please either point out where it is used, modify some hard-coded things to use this variable instead, or remove this variable completely.

ExcChua replied:
Used in fetching the BackupLambda from Github release. Fixed.

Lsquared13 commented on terraform/modules/quorum-cluster/variables.tf:
```diff
@@ -460,3 +476,43 @@ variable "observer_node_counts" {
+  description = "Full URL path to Backup Lambda binary"
+}
+
+variable "backup_lambda_binary_path" {
```
I'm searching through the PR and I can't find where this `backup_lambda_binary_path` variable is used. Please either point out where it is used, modify some hard-coded things to use this variable instead, or remove this variable completely.

Lsquared13 commented on terraform/variables.tf:
```diff
@@ -460,3 +460,66 @@ variable "observer_node_counts" {
+variable "backup_interval" {
+    default     = ""
+    description = <<DESCRIPTION
+Schedule expression for backup event, see https://docs.aws.amazon.com/lambda/latest/dg/tutorial-scheduled-events-schedule-expressions.html for examples
```
Thanks for adding this link. I was going to ask you to do that but you beat me to it.

ExcChua committed:
Updates as suggested by Louis

Updates as suggested by Louis

ExcChua committed:
Removed duplicate backup_lambda_binary definition

Removed duplicate backup_lambda_binary definition, changed uppercase to lowercase in local-exec.

ExcChua committed:
Fixed wrong name for backup_lambda_output_path

Fixed wrong name for backup_lambda_output_path

ExcChua committed:
Added missing var

Added missing var

ExcChua committed:
Updated security group and moved rules out

Updated security group and moved rules out

ExcChua committed:
Removed the backup_lambda_binary_path

Removed the backup_lambda_binary_path, changed security group id

ExcChua commented:
I think I've made all required changes as you've noted.

ExcChua committed:
Fix security rule for backup lambda security group

Fix security rule for backup lambda security group

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,352 @@
+  count  = "${signum(lookup(var.maker_node_counts, var.aws_region, 0))}"
+
+  vpc_id = "${aws_vpc.quorum_cluster.id}"
+  tags {
```
style nit: Add empty line before `tags`

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,352 @@
+  tags {
+     Name = "BackupLambdaSSH-${var.network_id}-${var.aws_region}-RouteTable"
+  }
+  route {
```
style nit: Add empty line before `route`

Lsquared13 commented on terraform/modules/quorum-cluster/variables.tf:
```diff
@@ -460,3 +476,38 @@ variable "observer_node_counts" {
+
+# Backup Lambda related variables
+variable "backup_interval" {
+  description = "Schedule expression for backup"
```
nit: Put the description you wrote elsewhere here as well

Lsquared13 commented on terraform/modules/quorum-vpc-peering/variables.tf:
```diff
@@ -31,3 +31,8 @@ variable "quorum_vpc_peering_counts" {
   type        = "map"
 }
+
+variable "quorum_vpc_main_route_table" {
```
I don't see this being specified anywhere. Is this variable used?

ExcChua replied:
Yes.

Lsquared13 commented on terraform/variables.tf:
```diff
@@ -93,7 +93,7 @@ variable "generate_metrics" {
 variable "create_alarms" {
   description = "Whether or not to create CloudWatch alarms. They will not function if generate_metrics is false."
-  default     = false
+  default     = true
```
It seems like you changed defaults to help with testing in this file. Please don't change default values unnecessarily.

ExcChua replied:
In order to investigate cloudwatch metrics and alarms, and due to the way Terraform Enterprise works, I have to change defaults. If you have a better way, please provide instructions.

Lsquared13 replied:
Why can't you just set the variables in `terraform.tfvars` (Terraform) or the Variables tab (Terraform Enterprise)?

You're free to change the defaults for testing purposes but these variables will need to be changed back before I pull this in

Lsquared13 commented on .gitignore:
```diff
@@ -8,3 +8,46 @@ terraform/cert-tool/terraform.tfstate*
+terraform/us-east-2/BackupLambda
+terraform/us-west-1/BackupLambda
+terraform/us-west-2/BackupLambda
+workingterraformvars.txt
```
This looks like something you added for testing. Remove it please.

ExcChua replied:
Remove which? Please be specific.

Lsquared13 replied:
This refers to the line I made the comment on: `workingterraformvars.txt`

Lsquared13 commented on .gitignore:
```diff
@@ -8,3 +8,46 @@ terraform/cert-tool/terraform.tfstate*
 terraform/cert-tool/.terraform*
 
 .vscode/
+terraform/eu-west-2-BackupLambda.zip
```
So if you use this repository after a fresh checkout, where do things end up being downloaded?

There's 3 things per region here, are they all used?  This is a nit, but it might be preferable to put the things that are downloaded into a particular directory to avoid cluttering the workspace.

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,352 @@
+  count               = "${signum(lookup(var.maker_node_counts, var.aws_region, 0))}"
+
+  name_prefix         = "BackupLambda-${var.network_id}-${var.aws_region}-" 
+  schedule_expression = "${var.backup_interval}"
```
`backup_interval` currently has no default. What happens if you run it with default values? (e.g. empty string for `backup_interval`)

I think ideally this backup stuff would be optional and not setting `backup_interval` would disable all of it. I won't insist on that if you think it's more work to handle than it's worth, but it should definitely do something reasonable if I use default values.

Lsquared13 commented:
Also please rebase this when you get the chance. Make sure your PR is up to date and ready for merging. We'll probably only need another exchange or two before this is ready.

ExcChua committed:
Fix styles and output

Fix styles and output

ExcChua committed:
Fixed dependency so that count can be computed

Fixed dependency so that count can be computed

ExcChua committed:
Fixed prefix for dependency

Fixed prefix for dependency

ExcChua committed:
Remove instance_tags

Remove instance_tags

ExcChua committed:
Fix indexing of ID

Fix indexing of ID

ExcChua committed:
Replace length with local

Replace length with local

ExcChua committed:
Changed count to index using maker_node_count

Changed count to index using maker_node_count

ExcChua committed:
Changed  count to lookup of maker_node_counts

Changed  count to lookup of maker_node_counts

ExcChua committed:
Fixed observer and validator outputs

Fixed observer and validator outputs

ExcChua committed:
Revert "Fix security rule for backup lambda security group"

This reverts commit 11df9559b22adc2b040b96c115b552bc9f9b1ca9.

ExcChua committed:
Fix output node

Fix output node

ExcChua committed:
Update output attribute

Update output attribute

ExcChua committed:
Update outputs

Update outputs

ExcChua committed:
Fix validator and observer output

Fix validator and observer output

ExcChua committed:
Fix outputs

Fix outputs

ExcChua committed:
Fix output

Fix output

ExcChua committed:
Fix validator, observer output

Fix validator, observer output

ExcChua committed:
Update quorum for output fix

Update quorum for output fix

ExcChua committed:
Reverted quorum

Reverted quorum

ExcChua committed:
Fixed retrieval of quorum_maker_node

Fixed retrieval of quorum_maker_node

ExcChua committed:
Update quorum to query current maker instance nodes

Update quorum to query current maker instance nodes

ExcChua committed:
Added depends_on to quorum_maker_dns aws_instances

Added depends_on to quorum_maker_dns aws_instances in order to delay the querying

ExcChua committed:
Trying a new syntax

ExcChua committed:
Added a new dependency for quorum_maker

ExcChua committed:
Fixed how aws_instance quorum_maker_node is created

ExcChua committed:
Fixed variable name in lookup

ExcChua committed:
Update count to remove query if it doesn't exist

ExcChua committed:
Fix how quorum validator is looked up

ExcChua committed:
Fixed validator and observer lookups

Lsquared13 commented on .gitignore:
```diff
@@ -8,3 +8,4 @@ terraform/cert-tool/terraform.tfstate*
 terraform/cert-tool/.terraform*
 
 .vscode/
+terraform/temp
```
I'd recommend doing something like what I did with the .gitignore [here](https://github.com/Eximchain/terraform-aws-quorum-cluster/blob/master/terraform/modules/quorum-vault/certs/.gitignore).

Putting that file in a directory allows you to check in the directory but keep it empty. It basically says "Ignore everything except .gitignore itself".

ExcChua committed:
Rolled back example, and updated variables

ExcChua committed:
Merge pull request #1 from Eximchain/master

Update fork

ExcChua committed:
Update security rule to allow all outgoing for BackupLambda

ExcChua committed:
Added feature to disable backup

Added feature to disable backup

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,355 @@
+}
+
+resource "aws_sns_topic" "backup_event" {
+  count       = "${var.backup_enabled ? signum(lookup(var.maker_node_counts, var.aws_region, 0)) : 0}"
```
Don't you still have to set these resources to be created if there are any nodes, rather than just any maker nodes?

I recall us deciding to do backups on validators, makers, and observers in that order of preference, so we need these resources if there are any nodes, right?

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,355 @@
+  provisioner "local-exec" {
+     command = <<EOT
+if [ ! -e ${var.aws_region}-${var.backup_lambda_output_path} ]; then 
+   wget -O ${var.aws_region}-${var.backup_lambda_output_path} ${var.backup_lambda_binary_url}
```
Should this be going into `terraform/temp`?

ExcChua replied:
Fixed

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,355 @@
+  count       = "${var.backup_enabled ? signum(lookup(var.maker_node_counts, var.aws_region, 0)) : 0}"
+
+  description = "Used for encrypting SSH keys on S3"
+  tags {
```
style nit: add empty line above `tags`

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,355 @@
+  vpc_id             = "${aws_vpc.quorum_cluster.id}"
+  availability_zone  = "${lookup(var.az_override, var.aws_region, "") == "" ? element(data.aws_availability_zones.available.names, count.index) : element(split(",", lookup(var.az_override, var.aws_region, "")), count.index)}"
+  cidr_block         = "${cidrsubnet(data.template_file.quorum_maker_cidr_block_lambda.rendered, 3, count.index)}"
+  tags {
```
style nit: add empty line above `tags`

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,355 @@
+  count      = "${var.backup_enabled ? signum(lookup(var.maker_node_counts, var.aws_region, 0)) : 0}"
+
+  vpc        = true
+  tags {
```
style nit: add empty lines above `tags`

ExcChua replied:
Fixed

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,355 @@
+resource "aws_eip" "gateway_ip" {
+  count      = "${var.backup_enabled ? signum(lookup(var.maker_node_counts, var.aws_region, 0)) : 0}"
+
+  vpc        = true
```
Style nit: remove extra spaces after `vpc` as so: `vpc = true`

ExcChua replied:
Fixed

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,355 @@
+
+  allocation_id = "${aws_eip.gateway_ip.0.id}"
+  subnet_id     = "${aws_subnet.quorum_maker.0.id}"
+  tags {
```
style nit: add empty line above `tags`

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,355 @@
+resource "aws_nat_gateway" "backup_lambda" {
+  count         = "${var.backup_enabled ? signum(lookup(var.maker_node_counts, var.aws_region, 0)) : 0}"
+
+  allocation_id = "${aws_eip.gateway_ip.0.id}"
```
style nit: prefer `${element(aws_eip.gateway_ip.*.id, 0)}` over `${aws_eip.gateway_ip.0.id}`

ExcChua replied:
${element...} is harder to read than .0.id so not going to change this.

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,355 @@
+  count         = "${var.backup_enabled ? signum(lookup(var.maker_node_counts, var.aws_region, 0)) : 0}"
+
+  allocation_id = "${aws_eip.gateway_ip.0.id}"
+  subnet_id     = "${aws_subnet.quorum_maker.0.id}"
```
style nit: prefer `${element(aws_subnet.quorum_maker.*.id, 0)}` over `${aws_subnet.quorum_maker.0.id}`

ExcChua replied:
${element...} is harder to read than .0.id

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,355 @@
+  count  = "${var.backup_enabled ? signum(lookup(var.maker_node_counts, var.aws_region, 0)) : 0}"
+
+  vpc_id = "${aws_vpc.quorum_cluster.id}"
+  tags {
```
style nit: add empty line above `tags`

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,355 @@
+  tags {
+     Name = "BackupLambdaSSH-${var.network_id}-${var.aws_region}-RouteTable"
+  }
+  route {
```
style nit: add empty line above `route`

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,355 @@
+  }
+  route {
+    cidr_block     = "0.0.0.0/0"
+    nat_gateway_id = "${aws_nat_gateway.backup_lambda.0.id}"
```
style nit: prefer `${element(aws_nat_gateway.backup_lambda.*.id, 0)}` over `${aws_nat_gateway.backup_lambda.0.id}`

ExcChua replied:
${element....} is harder to read then .0.id, so not going to change this.

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum.tf:
```diff
@@ -496,39 +496,69 @@ data "aws_ami" "quorum" {
 # ---------------------------------------------------------------------------------------------------------------------
-data "aws_instance" "quorum_maker_node" {
-  count = "${aws_autoscaling_group.quorum_maker.count}"
+data "aws_instances" "quorum_maker_dns" {
```
Why introduce `aws_instances` here? What do you get out of it that you can't get out of individual `aws_instance` data sources?  I don't mind the idea of replacing multiple `aws_instance` data sources with a single `aws_instance`, but I strongly prefer not to include both.

This comment applies to other node types as well

ExcChua replied:
aws_instances is introduced because your final list of output (maker_node, observer_node, validator_node) is empty after first runs.

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum.tf:
```diff
@@ -496,39 +496,69 @@ data "aws_ami" "quorum" {
-  filter {
-    name   = "tag:aws:autoscaling:groupName"
-    values = ["${element(aws_autoscaling_group.quorum_maker.*.name, count.index)}"]
+  instance_tags {
```
Can we do better than just relying on the `Name` tag here? Using tags like `NetworkId` and `Role` is probably less fragile.

This comment applies to other node types as well

Lsquared13 replied:
Bump this comment. Using the `Name` tag seems fragile. If you think it's better to use that than to use a combination of other tags please explain why.

ExcChua replied:
Can't locate where this is, but please explain fragile how?
Unless someone else, either you or me is changing code at the last minute, I'm not going to change this now.

Lsquared13 commented on terraform/modules/quorum-cluster-region/variables.tf:
```diff
@@ -387,3 +401,67 @@ variable "observer_node_counts" {
+# this is the lambda zip, must be a relative path
+# eg "BackupLambda.zip"
+variable "backup_lambda_output_path" {
+    default = "BackupLambda.zip"
```
Style nit: put `description` above `default`

Lsquared13 commented on terraform/modules/quorum-cluster-region/variables.tf:
```diff
@@ -387,3 +401,67 @@ variable "observer_node_counts" {
+# This is the full path to the private key 
+# eg "/Users/xxxx/.ssh/cert"
+variable "private_key_path" {
+    default = ""
```
Style nit: put `description` above `default`

Lsquared13 commented on terraform/modules/quorum-cluster-region/variables.tf:
```diff
@@ -387,3 +401,67 @@ variable "observer_node_counts" {
+# This is the public key path
+# eg "/Users/xxxx/.ssh/cert.pub"
+variable "public_key_path" {
+    default = ""
```
Style nit: put `description` above `default`

Lsquared13 commented on terraform/modules/quorum-cluster/variables.tf:
```diff
@@ -465,3 +481,45 @@ variable "observer_node_counts" {
+}
+
+variable "backup_lambda_binary" {
+  default = ""
```
Style nit: put `description` above `default`

Lsquared13 commented on terraform/modules/quorum-cluster/variables.tf:
```diff
@@ -465,3 +481,45 @@ variable "observer_node_counts" {
+}
+
+variable "backup_lambda_binary_url" {
+  default = ""
```
Style nit: put `description` above `default`

Lsquared13 commented on terraform/variables.tf:
```diff
@@ -465,3 +465,70 @@ variable "observer_node_counts" {
+
+
+variable "backup_lambda_binary" {
+  default = "BackupLambda"
```
Style nit: put `description` above `default`

Lsquared13 reviewed (changes requested):
Mostly style nits and a few more questions. Please make sure you address all the comments in some way. This should be ready to pull in any revision now.

Additionally you should be aware of and fix the error coming from terraform enterprise (may be related to my question about the file). See attached screenshot.

<img width="1136" alt="screen shot 2018-11-30 at 10 13 56 am" src="https://user-images.githubusercontent.com/4107447/49297479-df119080-f488-11e8-9b50-26e9557519de.png">

ExcChua committed:
Fix for temp directory and backup all nodes

ExcChua committed:
Create temp directory, remove local_file.EncryptedSSHKey

ExcChua committed:
Updated aws_nat_gateway to use system subnet

ExcChua committed:
Create and merge subnets

ExcChua committed:
Update backup_lambda to be able to access all subnets

ExcChua committed:
Removed system subnet and fix NAT gateway subnet

ExcChua committed:
Attempt to fix backup_lambda for aws_subnet.* not found issue

ExcChua committed:
Add depends_on for aws_nat_gateway

ExcChua committed:
Edited aws_subnet.*

ExcChua committed:
Update references to aws_subnet.*

ExcChua committed:
Update aws_nat_gateway subnet_id

ExcChua committed:
Retry

ExcChua committed:
update aws_nat_gateway subnet

ExcChua committed:
Fix NAT gateway for 3 node types

ExcChua committed:
Add routing table, routing entries and subnet debugging

ExcChua committed:
Added quorum_backup_lambda_cidr_block

ExcChua committed:
Fix quorum_backup_lambda_cidr_block subnet prefix

ExcChua committed:
Fix subnet_id in aws_instance

ExcChua committed:
Fix subnet_id in aws_route_table_association

ExcChua committed:
Fix subnet_id in aws_instance for maker

ExcChua committed:
Update subnet CIDR prefix

ExcChua committed:
Add SSH incoming for debugging

ExcChua committed:
Update security group

ExcChua committed:
update security_group names

ExcChua committed:
Add instances for debugging

ExcChua committed:
update conditions

ExcChua committed:
Added prefix

ExcChua committed:
Added prefix

ExcChua committed:
add eip

ExcChua committed:
set associate_public_ip_address = true in instances

ExcChua committed:
Update security group and rules into one

ExcChua committed:
Removed debugging aws_instance definitions

ExcChua committed:
Updated subnet_ids for lambda function

ExcChua committed:
Updated subnet_ids for backup_lambda

ExcChua committed:
update subnet_ids

ExcChua committed:
Add public, private subnet

ExcChua committed:
Recheck-in to trigger Terraform build

ExcChua committed:
Removed public/private subnet

ExcChua committed:
Update private/public subnet

ExcChua committed:
Test

ExcChua committed:
Fix heredoc issue

ExcChua committed:
Revised backup_lambda subnets

ExcChua committed:
Fix missing ?

ExcChua committed:
Retry

ExcChua committed:
Fix

ExcChua committed:
Merge

ExcChua committed:
Fix public/private subnets

ExcChua committed:
Update public/private subnets

ExcChua committed:
subne tinkling

ExcChua committed:
Fixed private/public subnet in lambda function

ExcChua committed:
Fixed count for private/public subnet

ExcChua committed:
Add routing table and associations for public/private subnet

ExcChua committed:
Add vpc_id for private/public route table

ExcChua committed:
Removed routing table for all quorum_* nodes

ExcChua committed:
Remove routing table association for all quorum nodes

ExcChua committed:
Removed routing tables and associations

ExcChua committed:
Removed debug instance declarations

ExcChua committed:
Updated signum lookup to local var

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,354 @@
+  filename = "${var.backup_lambda_ssh_private_key_path}"
+}
+
+locals {
```
Wow this is a cool feature!

style nit: move this above the `backup_lambda_ssh_private_key` data source, just under the providers

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,354 @@
+resource "aws_iam_role" "backup_lambda" {
+  count = "${var.backup_enabled ? local.signum_lookup : 0}"
+
+  name  = "iam_for_backup_lambda-${var.network_id}-${var.aws_region}"
```
style nit: use kebab case consistently here

change to `iam-role-for-backup-lambda-${var.network_id}-${var.aws_region}`

ExcChua replied:
Fixed.

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,354 @@
+EOF
+}
+
+resource "null_resource" "mkdir_temp" {
```
Here you use the directory `${path.module}/tmp/` but in your .gitignore, you ignore the directory `terraform/temp`.  I'm guessing these are supposed to agree with each other. If so pick either `temp` or `tmp` and use it consistently. If they are supposed to be unrelated perhaps change the name of one so they don't seem like the same thing modulo a typo.

Lsquared13 replied:
Bump this comment.

It seems like `${path.module}/tmp/` used here doesn't agree with `terraform/temp` listed in `.gitignore`

Should you replace with local `temp_lambda_zip_path`?

This comment applies for the whole resource

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,354 @@
+}
+
+locals {
+  temp_lambda_zip_path = "${path.module}/tmp/${var.aws_region}-${var.backup_lambda_output_path}"
```
same `tmp` vs `temp` issue as above

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,354 @@
+   policy_arn = "${aws_iam_policy.allow_backup_lambda_logging.arn}"
+}
+
+locals {
```
style nit: put `locals` up at the top of the file just under any providers

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,354 @@
+  description = "Used for encrypting SSH keys on S3"
+
+  tags {
+     name = "BackupLambda-${var.network_id}-${var.aws_region}-KMS"
```
nit: I believe this tag should be `Name` rather than `name`

I'm not sure if it's case sensitive for console display purposes

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,354 @@
+resource "aws_security_group" "allow_all_for_backup_lambda" {
+  count       = "${var.backup_enabled ? local.signum_lookup : 0}"
+
+  name        = "BackupLambdaSSH-${var.network_id}-${var.aws_region}-allow_all_outgoing"
```
style nit: use kebab case consistently

Change to `BackupLambdaSSH-${var.network_id}-${var.aws_region}-allow-all-outgoing`

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum.tf:
```diff
@@ -81,13 +81,90 @@ resource "aws_subnet" "quorum_observer" {
+
+  tags {
+    Name      = "quorum-network-${var.network_id}-Public-Subnet"
+    NodeType  = "Public-Subnet"
```
I don't think this is an appropriate tag, given that it's a subnet and not a node.

If you think it's helpful perhaps rename it?

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum.tf:
```diff
@@ -81,13 +81,90 @@ resource "aws_subnet" "quorum_observer" {
+
+  tags {
+    Name      = "quorum-network-${var.network_id}-Private-Subnet"
+    NodeType  = "Private-Subnet"
```
Same as above comment, I think this `NodeType` tag should have a different name on subnets.

Lsquared13 commented on terraform/modules/quorum-cluster-region/variables.tf:
```diff
@@ -387,3 +401,67 @@ variable "observer_node_counts" {
+  description = <<DESCRIPTION
+URL to retrieve the backup lambda binary.
+DESCRIPTION
+  default     = "https://github.com/EximChua/BackupLambda/releases/download/0.1/BackupLambda.zip"
```
Is `BackupLambda` hosted by the `Eximchain` organization yet?

If not, we should put that in our organization and point this at `https://github.com/Eximchain/BackupLambda/releases/download/0.1/BackupLambda.zip` by default instead

ExcChua replied:
There are access issues with Eximchain/BackupLambda for some unknown reason.
If you have time, please investigate, or delete the entire BackupLambda, so I can recreate it and see if that resolves the issue.

Lsquared13 replied:
I checked and it says you are Admin. What can't you do and what errors do you get when you try to do it? If you can't interact with it via CLI, are you sure you have an SSH key uploaded to GitHub in your SSH agent?

Follow up with me on this via email if you need more help.

Lsquared13 commented on terraform/variables.tf:
```diff
@@ -465,3 +465,70 @@ variable "observer_node_counts" {
+  description = <<DESCRIPTION
+URL to retrieve the Backup Lambda binary from.
+DESCRIPTION
+  default    = "https://github.com/EximChua/BackupLambda/releases/download/0.1/BackupLambda.zip"
```
Same as comment on other variables file, ideally this would point to an eximchain repository.

Lsquared13 reviewed (changes requested):
Mostly style comments at this point and a few outstanding questions. I don't think anything I've asked requires any sort of major rework, please feel free to push back if you think anything I've asked will take more than a couple days to respond to.

ExcChua committed:
Fix nits

Lsquared13 commented on terraform/modules/quorum-cluster-region/quorum-cluster-region-backup-event.tf:
```diff
@@ -0,0 +1,350 @@
+  vpc_id      = "${aws_vpc.quorum_cluster.id}"
+
+  tags {
+     name = "BackupLambda-${var.network_id}-${var.aws_region}-SG"
```
nit: I believe the `name` tag should be capitalized as `Name` to work with the AWS console

ExcChua replied:
Fixed.

Lsquared13 commented on terraform/modules/quorum-cluster-region/variables.tf:
```diff
@@ -387,3 +401,67 @@ variable "observer_node_counts" {
+  description = <<DESCRIPTION
+Enable backup of chain data.
+DESCRIPTION
+  default = "true"
```
nit: default to `false` to make our lives easier when testing

ExcChua replied:
Fixed

Lsquared13 commented on terraform/modules/quorum-cluster/variables.tf:
```diff
@@ -465,3 +481,45 @@ variable "observer_node_counts" {
+  description = <<DESCRIPTION
+Enable backup of chain data.
+DESCRIPTION
+  default = "true"
```
nit: default to false

ExcChua replied:
Fixed

Lsquared13 commented on terraform/variables.tf:
```diff
@@ -465,3 +465,70 @@ variable "observer_node_counts" {
+  description = <<DESCRIPTION
+Enable backup of chain data.
+DESCRIPTION
+  default = "true"
```
nit: default to false

ExcChua replied:
Fixed

Lsquared13 reviewed (changes requested):
I believe this is ready to pull in once the few outstanding comments are addressed. Please confirm that you've tested the latest version both with and without backup enabled

ExcChua committed:
Fix nits

ExcChua committed:
Fix aws_s3_bucket quorum_backup

Reference to quorum_backup from aws_iam_policy.quorum

ExcChua committed:
Removed ternary expression

ExcChua committed:
Fix temp directory

ExcChua committed:
Fix tempdir

ExcChua committed:
Fix local

ExcChua committed:
Added configurable SSH user/pass

ExcChua committed:
Fixed missing ssh user/pass in quorum-cluster main and vars
